asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,758 @@
+        power_per_heatpump: Power consumption per heat pump in kW.
+        voltage_factor: Factor to account for voltage drop (default 0.95).
+
+    Returns:
+        DataFrame with heat pump installation capacity assessment for each LSOA.
+    """
+    # Convert headroom from MVA to kW
+    lsoa_data["headroom_kw"] = lsoa_data["distributed_headroom"] * 1000 * voltage_factor
+
+    # Calculate maximum number of heat pumps that could be installed
+    lsoa_data["max_heatpumps"] = np.floor(lsoa_data["headroom_kw"] / power_per_heatpump)
+
+    # Calculate percentage of households that could install heat pumps
+    lsoa_data["heatpump_installation_percentage"] = (
+        lsoa_data["max_heatpumps"] / lsoa_data["household_count"] * 100
+    ).clip(upper=100, lower=0)
+
+    # Calculate excess or deficit capacity
+    lsoa_data["capacity_difference_kw"] = lsoa_data["headroom_kw"] - (
+        lsoa_data["household_count"] * power_per_heatpump
+    )
+
+    return lsoa_data
+
+
+def calculate_grid_capacity() -> pd.DataFrame:
+    """
+    Calculate the grid capacity for heat pump installations across all LSOAs.
+
+    This function performs the following steps:
+    1. Generate and process substation data from all grid operators
+    2. Load and process LSOA and household data
+    3. Distribute substation headroom to LSOAs
+    4. Assess heat pump suitability based on available capacity
+
+    Returns:
+        DataFrame with grid capacity assessment results for each LSOA
+    """
+    # Generate and process substation data
+    substations = generate_substations_df()
+    substations["firm_capacity_mva"] = substations["firm_capacity_mva"].apply(
+        _parse_capacity
+    )
+    substations["peak_demand_mva"] = substations["peak_demand_mva"].apply(
+        _parse_capacity
+    )
+
+    # Apply substation rating scaling factor
+    substations["firm_capacity_mva"] = (
+        substations["firm_capacity_mva"] * SUBSTATION_SCALING_FACTOR
+    )
+
+    # Calculate headroom per substation
+    substations["headroom_mva"] = (
+        substations["firm_capacity_mva"] - substations["peak_demand_mva"]
+    )
+
+    # Load and process LSOA boundary data
+    lsoa_boundaries = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/Output_Areas_2021_EW_BGC_V2_4299916833741807639.geojson",
+    )
+    lsoa_gdf = aggregate_oa_to_lsoa(lsoa_boundaries)
+
+    # Load and process household data
+    households_df = get_datasets.get_df_ons_number_of_households()
+    households_df = households_df.rename(
+        mapping={"mnemonic": "LSOA21CD", "2021": "household_count"}
+    ).to_pandas()
+
+    # Process substation and LSOA data
+    headroom_df = process_substation_lsoa_data(substations, lsoa_gdf, households_df)
+
+    # Assess heat pump suitability
+    consumption_per_heatpump = POWER_PER_HEATPUMP / COEFFICIENT_OF_PERFORMANCE
+    result = assess_heatpump_suitability(
+        headroom_df, consumption_per_heatpump, TAKEUP_RATE
+    )
```

Comment:
It doesn't look like this argument should be TAKEUP_RATE as the function says its voltage_factor?